File: Programmers/pro_42888_오픈채팅방.py
from collections import defaultdict
def solution(record):
    name_dict = defaultdict(str)
    
    #change name
    for i in record:
        if(i[0]=='E' or i[0]=='C'):
            cmd = i.split(' ')
            name_dict[cmd[1]] = cmd[2]
    
    answer = []
    for i in record:
        cmd = i.split(' ')
        if(cmd[0]=='Enter'):
            answer.append(name_dict[cmd[1]]+"님이 들어왔습니다.")
        elif(cmd[0]=='Leave'):
            answer.append(name_dict[cmd[1]]+"님이 나갔습니다.")
        elif(cmd[0]=='Change'):
            continue    
    return answer

File: Programmers/pro_42888_오픈채팅방.js
function solution(record) {
    const len = record.length
    const name_dict = []
    for (let i = 0; i < len; i++) {
        if (record[i][0] !== "L") {
            let [_, id, name] = record[i].split(" ")
            name_dict[id] = name
        }
    }
    let answer = [];
    for (let i = 0; i < len; i++) {
        let firstChar = record[i][0]
        if (firstChar !== "C") {
            let [_, id, name] = record[i].split(" ")
            if (firstChar === "E") {
                answer.push(name_dict[id] + "님이 들어왔습니다.")
            } else if (firstChar === "L") {
                answer.push(name_dict[id] + "님이 나갔습니다.")
            }
        }
    }
    return answer;
}